contracts/hackatom/src/contract.rs

```diff
@@ -41,21 +43,31 @@ pub fn raw_query(key: &[u8]) -> Result<Vec<u8>> {
 
 pub static CONFIG_KEY: &[u8] = b"config";
 
-pub fn init<T: Storage>(store: &mut T, params: Params, msg: Vec<u8>) -> Result<Response> {
+pub fn init<T: Storage, U: Addresser>(
+    store: &mut T,
+    addr: &U,
```

Comment:
At this point we might want to inject addresser as part of a bigger API in a single `api` argument. `api.address.canonicalize` could then do the job and can be extended to fun stuff like e.g. `api.crypto.scrypt`. I see this API argument as gobally unique and immutable, so `store` should remain separate.

Reply:
Great. That mirrors my thinking last night after pushing this code.

Storage has state reference and is separate.

I would rename this to Precompiles and then pass in pure functions with no state access. We can add more there later. Maybe we even make them optional, so contracts that do not need one will not fail if it is missing.

Querier also needs state, but different than the Storage. I think to make this a third generic argument that can be mocked out.

Of course this all needs some solid documentation. But defining the proper abstractions in code helps provide order to the docs as well